Pull out a scaled slider class for finer grained control of slice positions

Andreas-Forster opened the pull request:
This PR fixes #50 by introducing a new class [SubDividedSliderAdapter](https://github.com/Andreas-Forster/scalismo-ui/blob/098d92cfd6bf82fb3bdbe9c4ccc46412a0a6ecb0/src/main/scala/scalismo/ui/view/util/SubDividedSliderAdapter.scala#L22). The sliders have by default a 100x finer resolution than now. This should fix for most use cases the issue #50 where the slices could not be made visible by using the sliders.

Andreas-Forster committed:
pull out a scaled slider for fine grained control of slice positions

Andreas-Forster commented:
Yet, this PR should serve as a discussion basis and testing is still needed before merging it.

Andreas-Forster committed:
cleaned interface of class SubDividedSliderAdapter

clangguth commented on src/main/scala/scalismo/ui/view/ViewportPanel.scala:
```diff
@@ -121,29 +121,23 @@ class ViewportPanel3D(frame: ScalismoFrame, override val name: String = "3D") ex
   override def name: String = axis.toString
 
-  private lazy val positionSlider = new Slider {
+  private val SubDivisions: Double = 100.0
```
This is obsolete and can be removed.

Andreas-Forster replied:
agree - addressed in new commit

clangguth commented on src/main/scala/scalismo/ui/view/ViewportPanel.scala:
```diff
@@ -121,29 +121,23 @@ class ViewportPanel3D(frame: ScalismoFrame, override val name: String = "3D") ex
-        positionSlider.value = positionSlider.value + 1
-      }
-    }
+    override def apply(): Unit = positionSlider.up
```
```suggestion
    override def apply(): Unit = positionSlider.up()
```

Andreas-Forster replied:
agree - addressed in new commit

clangguth commented on src/main/scala/scalismo/ui/view/ViewportPanel.scala:
```diff
@@ -121,29 +121,23 @@ class ViewportPanel3D(frame: ScalismoFrame, override val name: String = "3D") ex
-        positionSlider.value = positionSlider.value - 1
-      }
-    }
+    override def apply(): Unit = positionSlider.down
```
```suggestion
    override def apply(): Unit = positionSlider.down()
```

Andreas-Forster replied:
agree - addressed in new commit

clangguth commented on src/main/scala/scalismo/ui/view/properties/SlicingPositionPanel.scala:
```diff
@@ -44,38 +44,30 @@ class SlicingPositionPanel(override val frame: ScalismoFrame) extends BorderPane
-          slider.value = slider.value - 1
-        }
-      }
+      override def apply(): Unit = slider.down
```
```suggestion
      override def apply(): Unit = slider.down()
```

Andreas-Forster replied:
agree - addressed in new commit

clangguth commented on src/main/scala/scalismo/ui/view/properties/SlicingPositionPanel.scala:
```diff
@@ -44,38 +44,30 @@ class SlicingPositionPanel(override val frame: ScalismoFrame) extends BorderPane
-          slider.value = slider.value + 1
-        }
-      }
+      override def apply(): Unit = slider.up
```
```suggestion
      override def apply(): Unit = slider.up()
```

Andreas-Forster replied:
agree - addressed in new commit

clangguth commented on src/main/scala/scalismo/ui/view/util/SubDividedSliderAdapter.scala:
```diff
@@ -0,0 +1,44 @@
+
+class SubDividedSliderAdapter(private val _slider: Slider, private val factor: Double = 100.0) {
+
+  def slider: Slider = _slider
```
There is no need to introduce a private variable here, as it's an object anyway which is made accessible using this getter.

Therefore, my suggestion is to simply make `slider` a val. In the same vein, there's not really a need to keep `factor` private.

Andreas-Forster replied:
agree - addressed in new commit

clangguth reviewed (changes requested):
Looks good to me in general, I've added a few tiny nitpicking suggestions.

The only thing that still bothers me a bit is that most of the time references are made to the adapter, but sometimes to the contained slider. (listenTo(x.slider.slider)).

This could be overcome by making the adapter itself a ScalismoPublisher that listens to the slider events and re-publishes them on behalf of itself. Then the only occurence where direct access to the contained slider would still be needed is in the layout() calls, because they expect an actual Swing component.

Andreas-Forster committed:
made SubDividedSliderAdapter a ScalismoPublisher and introduced FancySliderDecorator

Andreas-Forster commented:
I made the SubDivdedSliderAdapter a ScalismoPublisher. Further, I had to introduce a new FancySliderDecorator as the FancySlider does not handle the position of the labels correctly.

It looks now that we have quite a few sliders classes. Not sure if there is a better solution.

clangguth commented:
Hmmm... ok, please give me a couple of days to take a closer look at how this can be properly solved.

clangguth commented:
Sorry, I don't currently find the time for a proper solution, but here's an idea/outline of a "one-size fits all" slider class which covers all existing use cases, maybe someone finds a way to implement it:
- ensure that it works with multiple types, not just integers. This would mean creating a new generic class which mimicks the interface of the existing Slider class, but in a typed way. Implementation-wise it could either wrap a JSlider or a Slider.
- include the formatting part of the FancySlider, so that labels can be shown if needed.
- the tricky part: ensure that it works for any domain (min/max) and that it allows to set a "step size" (or factor, or granularity, or whatever you want to call it).

Essentially, one needs to find functions to map from that "application domain" (AD) to a "slider domain" (SD) and back, where SD only consists of integers and the step size is always 1. However, min/max of SD could be adjusted if needed, but the largest possible range is (Int.MinValue, Int.MaxValue). The mapping between AD and SD would also need to be bijective, so that changes (movement of the slider) properly get propagated back to the application.

Andreas-Forster committed:
introduced TypedSlider to remove having a lot of different sliders

Andreas-Forster commented:
I added a quick draft for a `TypedSlider[A]` class which should work for `Int`, `Float` and `Double`. I replaced all sliders except the anonymous class using directly `scala.swing.Slider` [here](https://github.com/unibas-gravis/scalismo-ui/blob/020aae765941c77042a818f0d8a329df7132a13c/src/main/scala/scalismo/ui/view/properties/GaussianProcessCoefficientsPanel.scala#L59).

A quick test showed that it seems to work. However, I have not yet tested all classes I changed and maybe there is an unhandled corner case. For example, one could require that the step-size is always strictly positive. A lower bound for the step-size based on the min and max value could also make sense, in order to not prevent that the slider's internal value does not overflow. These should however not occur if you use the sliders in a reasonable range. Any opinion if such checks should be added? Do you see other problems?

Further the implicit could probably be removed if one creates a slider class per type (int, float, double). Any opinion here if this should be changed?

marcelluethi commented:
I lost track about the status of this PR. Is it okay to close it or are we close to something that can be merged?

Andreas-Forster commented:
I think we can close it for now. We could still pick it up later on.
